src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -930,68 +918,91 @@ private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo p
-                        {
-                            yield return type;
-                        }
+                        GetTypesOfMembers(t, memberName, members, ref maybeWantDefaultCtor, isInvokeMemberExpressionAst: false, inferredTypes);
                     }
                 }
             }
 
             if (pseudoBinding.BoundArguments.TryGetValue("Begin", out argument))
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
-                {
-                    yield return type;
-                }
+                GetInferredTypeFromScriptBlockParameter(argument, inferredTypes);
             }
 
             if (pseudoBinding.BoundArguments.TryGetValue("Process", out argument))
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
-                {
-                    yield return type;
-                }
+                GetInferredTypeFromScriptBlockParameter(argument, inferredTypes);
             }
 
             if (pseudoBinding.BoundArguments.TryGetValue("End", out argument))
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
-                {
-                    yield return type;
-                }
+                GetInferredTypeFromScriptBlockParameter(argument, inferredTypes);
             }
         }
 
-        private IEnumerable<PSTypeName> InferTypesFromWhereAndSortCommand(CommandAst commandAst)
+        private void InferTypesFromWhereAndSortCommand(CommandAst commandAst, List<PSTypeName> inferredTypes)
+        {
+            InferTypesFromPreviousCommand(commandAst, inferredTypes);
+        }
+
+        private void InferTypesFromPreviousCommand(CommandAst commandAst, List<PSTypeName> inferredTypes)
         {
-            var parentPipeline = commandAst.Parent as PipelineAst;
-            if (parentPipeline != null)
+            if (commandAst.Parent is PipelineAst parentPipeline)
             {
                 int i;
                 for (i = 0; i < parentPipeline.PipelineElements.Count; i++)
                 {
                     if (parentPipeline.PipelineElements[i] == commandAst)
+                    {
                         break;
+                    }
                 }
+
                 if (i > 0)
                 {
-                    foreach (var typename in InferTypes(parentPipeline.PipelineElements[i - 1]))
+                    inferredTypes.AddRange(InferTypes(parentPipeline.PipelineElements[i - 1]));
+                }
+            }
+        }
+
+        private void InferTypesFromSelectCommand(PseudoBindingInfo pseudoBinding, CommandAst commandAst, List<PSTypeName> inferredTypes)
+        {
+            if (pseudoBinding.BoundArguments.TryGetValue("Property", out _)
+               || pseudoBinding.BoundArguments.TryGetValue("ExcludeProperty", out _))
+            {
+                return;
```

Comment:
If the argument to `-Property` is a set of literal strings (most common case), then you could infer the "type" of the output object from them.

Reply:
Yes. I discussed this a bit with Jason. 
We don't currently have a good representation for this, but I would like to have that.
Same thing when you create a PSCustomObject from a hashtable.

But maybe something like "PSObject#Prop1/Prop2/Prop3"

It wouldn't say anything about the types of the properties, but that could maybe be encoded as well.